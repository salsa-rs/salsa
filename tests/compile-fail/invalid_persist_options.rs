#[salsa::input(persist)]
struct Input {
    text: String,
}

#[salsa::input(persist())]
struct Input2 {
    text: String,
}

#[salsa::input(persist(serialize = serde::Serialize::serialize))]
struct Input3 {
    text: String,
}

#[salsa::input(persist(deserialize = serde::Deserialize::deserialize))]
struct Input4 {
    text: String,
}

#[salsa::input(persist(serialize = serde::Serialize::serialize, deserialize = serde::Deserialize::deserialize))]
struct Input5 {
    text: String,
}

#[salsa::input(persist(serialize = serde::Serialize::serialize, serialize = serde::Serialize::serialize))]
struct InvalidInput {
    text: String,
}

#[salsa::input(persist(deserialize = serde::Deserialize::deserialize, deserialize = serde::Deserialize::deserialize))]
struct InvalidInput2 {
    text: String,
}

#[salsa::input(persist(not_an_option = std::convert::identity))]
struct InvalidInput3 {
    text: String,
}

#[salsa::tracked(persist)]
fn tracked_fn(db: &dyn salsa::Database, input: Input) -> String {
    input.text(db)
}

#[salsa::tracked(persist())]
fn tracked_fn2(db: &dyn salsa::Database, input: Input) -> String {
    input.text(db)
}

#[salsa::tracked(persist(serialize = serde::Serialize::serialize, deserialize = serde::Deserialize::deserialize))]
fn invalid_tracked_fn(db: &dyn salsa::Database, input: Input) -> String {
    input.text(db)
}

fn main() {}
