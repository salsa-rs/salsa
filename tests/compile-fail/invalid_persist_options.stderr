error: option `serialize` provided twice
  --> tests/compile-fail/invalid_persist_options.rs:26:65
   |
26 | #[salsa::input(persist(serialize = serde::Serialize::serialize, serialize = serde::Serialize::serialize))]
   |                                                                 ^^^^^^^^^

error: option `deserialize` provided twice
  --> tests/compile-fail/invalid_persist_options.rs:31:71
   |
31 | #[salsa::input(persist(deserialize = serde::Deserialize::deserialize, deserialize = serde::Deserialize::deserialize))]
   |                                                                       ^^^^^^^^^^^

error: unexpected argument
  --> tests/compile-fail/invalid_persist_options.rs:36:24
   |
36 | #[salsa::input(persist(not_an_option = std::convert::identity))]
   |                        ^^^^^^^^^^^^^

error: unexpected argument
  --> tests/compile-fail/invalid_persist_options.rs:51:26
   |
51 | #[salsa::tracked(persist(serialize = serde::Serialize::serialize, deserialize = serde::Deserialize::deserialize))]
   |                          ^^^^^^^^^
